### PR TITLE
fix(markdown): chat-scoped completeness — escaped chars, task lists, math/escape in `_remend`

### LINKED ISSUE #69: Revisit hand-rolled markdown parser: close setext / footnotes / HTML / indented-code gaps
**Coverage gap — documented non-parity with upstream remark.**

## Current state

`src/chat_sdk/shared/markdown_parser.py` is a ~880-line hand-rolled regex parser producing mdast-shaped output. It covers ATX headings, fenced code blocks, thematic breaks, blockquotes, ordered/unordered lists, GFM tables, inline emphasis/strong/delete/code/links/images, hard line breaks.

What it **doesn't** cover (listed at `docs/UPSTREAM_SYNC.md:442` as "by design, won't fix"):

- Setext headings (`=====` / `-----` underlines)
- Footnotes (`[^1]: ...`)
- Raw HTML nodes (block + inline)
- Indented code blocks (4-space indent)
- Escaped characters (`\*`, `\[`, `\\`, …)
- Backtick spans longer than 1 (`` ``code with ` inside`` ``)

The decision is recorded at `docs/DECISIONS.md:7-21`. Summary of the original rationale: (1) mdast compatibility — no Python parser produces mdast natively; (2) round-trip fidelity; (3) subset-is-sufficient; (4) zero runtime deps; (5) streaming renderer needs to understand the same constructs.

## Why reopen

Each "by design, won't fix" row accumulates risk:

- **LLM-generated content**: models frequently emit setext headings and escaped characters; these currently get rendered as raw text, breaking display on every adapter that pipes through our parser (all of them).
- **Upstream drift**: upstream remark handles these correctly. Every time upstream tests something that depends on them, we have to explicitly skip; the skip list grows.
- **Fragility**: the hand-rolled regex parser is already 880 lines and has had several correctness fixes for edge cases (see git history on `shared/markdown_parser.py`). Each new construct adds risk of subtle inline-span bugs; each fix risks regressing another construct.
- **Downstream asks**: consumers periodically ask why `\*not italic\*` doesn't render; the answer "we don't handle escaped characters" doesn't age well.

### Option A — close the gaps in the existing parser

Add setext, indented code, raw HTML, footnotes, escaped chars, and multi-backtick spans to `shared/markdown_parser.py`. Estimate: ~300–400 LOC, new edge cases to test. Keeps zero runtime deps.

### Option B — swap the backend

Evaluate `mistune` (has GFM + footnote plugins) or `markdown-it-py` (CommonMark + plugin ecosystem) with a thin translation layer to mdast. The original "no mdast from Python libs" objection stands, but the translation layer has shrunk in scope — we only need the node shapes we already emit (`paragraph`, `heading`, `code`, `list`, `listItem`, `blockquote`, `table`, `tableRow`, `tableCell`, `thematicBreak`, `emphasis`, `strong`, `delete`, `inlineCode`, `link`, `image`, `text`, `break`). A translator for those nodes is probably ~250 LOC. The parser itself shrinks to a call into the library.

Key concern with Option B: the `StreamingMarkdownRenderer` in `shared/streaming_markdown.py` needs to understand code-fence state and inline-marker state for mid-stream closure. If the parser library is used only for static parses (after-the-fact mdast), and streaming keeps its own state machine, they could diverge. Scope this carefully.

### Option C — close specific high-value gaps only

Order of ROI:
1. Escaped characters — common in LLM output; small implementation (~30 LOC).
2. Indented code blocks — common in technical content; small (~40 LOC).
3. Setext headings — common; medium (~60 LOC).
4. Multi-backtick spans — less common; small (~30 LOC).
5. Footnotes — rare in chat; defer.
6. Raw HTML — rare in chat; defer.

Option C is pragmatic: close items 1–3 in one PR, document 4–6 as remaining gaps, come back if consumers ask.

## Recommendation

Start with C. Benchmark against `mistune` in a separate spike to see whether the translator layer lands under 250 LOC; if yes, B becomes attractive for the 0.5 cycle.

## Acceptance

- [ ] Decide A / B / C (probably C, then reassess)
- [ ] If C: escaped chars + indented code + setext land with faithful tests from upstream remark test suite where available
- [ ] Remove matching rows from `docs/UPSTREAM_SYNC.md` non-parity table
- [ ] `docs/DECISIONS.md:7-21` updated to reflect the narrowed gap list
- [ ] Benchmark: parsing a 10KB mixed-content document stays under 5ms on CI hardware

### PR DESCRIPTION
**Pull request by patrick-chinchill**
## Summary

Chat-scoped completeness fixes for [issue #69](https://github.com/Chinchill-AI/chat-sdk-python/issues/69), targeting the gap subset that actually surfaces in LLM/agent output rather than chasing full CommonMark spec compliance.

**Sequenced after #99** (depends on `_is_excluded_asterisk` from that PR). Targets `main` for the 4.27 cycle — same release, separate PR for review clarity.

### Parser side (`markdown_parser.py`)

1. **Escaped characters** — inline regexes get `(?<!\\)` lookbehinds so `\*`, `\[`, `\_`, `\~~` stay literal; new `_unescape_punct` resolves the escape at text-leaf emission. Image alt text also unescapes. Inline-code contents deliberately preserve backslashes (CommonMark: code spans don't unescape).

2. **GFM task list checkboxes** — `- [ ]` / `- [x]` / `- [X]` parse to `listItem` with `checked: bool`. Ordered lists with `1. [x]` content are NOT coerced into tasks (task lists are bullet-only per GFM).

### Streaming side (`streaming_markdown.py::_remend`)

3. **Math regions** — new `_strip_math_regions` drops `$...$` (inline) and `$$...$$` (display) before the inline-marker counters, so `$a^* + b^*$` no longer accumulates phantom italic closers.

4. **Escape-aware counters** — new `_strip_escape_sequences` removes `\X` pairs from the counting buffer so the tilde/backtick/bracket counters in `_remend` no longer fire on escaped delimiters. The bracket walk's existing `\`-skip is preserved.

## What's deliberately NOT here

- **Word-internal asterisks (`5*3=15`)** — prototyped, reverted. Excluding word-internal `*` also breaks `text*foo*` (legitimate paired italic whose `*`s are also word-flanked). Distinguishing the two needs a CommonMark delimiter-stack algorithm. Documented as the next Option A item in `_is_excluded_asterisk`'s docstring.

- **Setext headings, footnotes, raw HTML, indented code blocks, definition lists** — analyzed in the parser-swap spike (PR #100). All silently dropped by the baseline and by every candidate without plugins. Not surfaced by LLM/chat output in practice ([analysis](https://github.com/Chinchill-AI/chat-sdk-python/blob/claude/markdown-parser-swap-spike/src/chat_sdk/shared/parser_spike/README.md)).

- **Multi-backtick code spans** — deferred. Less common than the items above and would require a non-trivial inline-parser change.

## Test plan

26 new tests across 3 classes:

- **`TestEscapedCharacters`** (8 tests): asterisk, underscore, bracket, tilde, backslash literal, mixed real/escaped emphasis, image-alt unescape, inline-code-preserves-backslash.
- **`TestTaskListItems`** (7 tests): unchecked, checked, uppercase X, plain item with no attr, mixed list state, marker stripped from content, ordered list not task-coerced.
- **`TestRemendChatCompleteness`** (11 tests): inline + display math regions, math doesn't break legitimate emphasis outside, math drops strike markers and bracket walk, escape-aware tilde/bracket/backtick counters, mixed escaped+real, escape neutrality.

Validation:

- [x] `uv run pytest tests/` — 3,621 pass, 1 pre-existing failure (`test_github_webhook.py::TestGitHubAdapterConstructor::test_throws_when_no_auth`, unrelated — same as #83/#99/#100)
- [x] `uv run ruff check src/ tests/ scripts/` — passes
- [x] `uv run ruff format --check src/ tests/ scripts/` — 193 files already formatted
- [x] `uv run python scripts/audit_test_quality.py` — 0 hard failures
- [x] No existing tests modified

## Scope notes

- **Dependency on #99**: this branch builds on `claude/wizardly-heisenberg-VLInV` and extends `_is_excluded_asterisk` added there. Should merge after #99.
- **No CHANGELOG entry** — same convention as the in-flight 4.27 port PRs (#84-#90 and #99); the consolidating sync PR (#83) owns release notes.
- **Doesn't close #69**: the broader parser-strategy question is now informed by PR #100's spike data plus this surgical fix. The remaining gaps (delimiter-stack for word-internal `*`, multi-backtick spans) can land in a follow-up if they ever materialize as user-visible bugs.

Related: #99 (the original three streaming bugs), #100 (the parser-swap spike that reframed the strategy decision).

---
_Generated by [Claude Code](https://claude.ai/code/session_013df8GRZykVd9Mg39ez1y8V)_

## Summary by CodeRabbit

* **New Features**
  * Support for GitHub-flavored task lists (checkbox state preserved, including empty items)

* **Bug Fixes**
  * Better handling of backslash-escaped punctuation and inline math so formatting is preserved
  * Improved nested-list and progressive table rendering to avoid misplaced markers and premature header emission
  * More faithful parse/stringify roundtrips to retain original markdown formatting

[Review Change Stack](https://app.coderabbit.ai/change-stack/Chinchill-AI/chat-sdk-python/pull/101?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: src/chat_sdk/shared/markdown_parser.py [hunk 1]
```diff
@@ -162,43 +169,159 @@ def get_node_value(node: Content) -> str:
...
+    ("strong_star", re.compile(r"(?<!\\)\*\*((?:[^\\]|\\.)+?)\*\*")),
     # Bold: __text__
-    ("strong_under", re.compile(r"__(.+?)__")),
+    ("strong_under", re.compile(r"(?<!\\)__((?:[^\\]|\\.)+?)__")),
     # Strikethrough: ~~text~~
-    ("delete", re.compile(r"~~(.+?)~~")),
+    ("delete", re.compile(r"(?<!\\)~~((?:[^\\]|\\.)+?)~~")),
     # Emphasis: *text*  (not preceded/followed by *)
-    ("emphasis_star", re.compile(r"(?<!\*)\*(?!\*)(.+?)(?<!\*)\*(?!\*)")),
+    ("emphasis_star", re.compile(r"(?<![*\\])\*(?!\*)((?:[^\\]|\\.)+?)(?<!\*)\*(?!\*)")),
```
[the remark this answers is not in the document]
Comment (patrick-chinchill) [reply src/chat_sdk/shared/markdown_parser.py]: Confirmed real bug — verified `\\*real italic*` parses as plain text instead of literal `\` + italic. Deferring the fix to a follow-up rather than blocking this PR. Reasoning:

The fix can't be done with the existing `(?<!\\)` lookbehind approach — Python's `re` doesn't support variable-width lookbehind, and any fixed-width alternation (e.g. `(?:(?<!\\)|(?<=\\\\))`) handles even-count backslashes but breaks 3+ backslashes (which should be escape-backslash + escape-delimiter = literal). The proper fix is a placeholder-based pre-pass: substitute every `\X` pair with a private-use Unicode placeholder before regex runs, then restore at text-leaf and inline-code emission. That's a moderate refactor of `_parse_inline`, `_unescape_punct`, and several inline regexes; safer as its own PR.

For now, added a comment block at `_INLINE_PATTERNS` (commit 9a2027f) documenting the limitation and the recommended fix path. The pattern is rare in LLM/chat output (which is this PR's scope) — backslash-emphasis sequences typically appear in technical docs explaining regex/glob syntax, which LLMs usually wrap in code fences. Will revisit when we have a concrete production case or when we tackle the broader parser-rewrite question from #100.

---
_Generated by [Claude Code](https://claude.ai/code)_

[the remark this answers is not in the document]
Comment (patrick-chinchill) [reply src/chat_sdk/shared/markdown_parser.py]: Update: actually fixed in `48fa4f3` rather than deferring. The placeholder-based pre-pass was smaller than I'd estimated (~50 LOC of helpers + ~10 LOC of wiring) and didn't break any existing tests.

Approach: `_protect_escapes` runs at the start of `_parse_inline` and replaces every `\X` escape with `<SENTINEL>X` (SENTINEL = U+0001, a control char that doesn't appear in chat markdown). The regex lookbehinds change from `(?<!\\)` to `(?<!\x01)`. Because the sentinel marks the escape unambiguously, any number of preceding backslashes is handled correctly:

- `\*` → `<S>*` — lookbehind sees `<S>`, skips. Correct (escaped).
- `\\*` → `<S>\*` — the `*` is preceded by `\` (not `<S>`), lookbehind passes, emphasis opens. Correct.
- `\\\*` → `<S>\<S>*` — the `*` is preceded by `<S>`, lookbehind skips. Correct (3 backslashes = escaped `\` + escaped `*`, both literal).

Two restore functions handle the emission modes:
- `_restore_escapes` (text leaves, image alt, link URL): `<S>X` → `X`
- `_restore_escapes_as_literal_pair` (inline code, where CommonMark preserves backslashes): `<S>X` → `\X`

Tests added: `test_escaped_backslash_followed_by_real_emphasis` pins your exact repro, plus a triple-backslash case for the harder counting. 3,636 / 3,637 total tests pass.

---
_Generated by [Claude Code](https://claude.ai/code)_
